File: app/src/main/java/com/ds/rickandmorty/services/CharacterServiceProvider.java
package com.ds.rickandmorty.services;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class CharacterServiceProvider {

    private static final String BASE_URL = "https://rickandmortyapi.com/api/character/";
    private static Retrofit retrofit;
    private static CharacterService characterService;
    
    public static CharacterService getInstance(){
        if(characterService == null){
            initializeRetrofit();
            characterService = retrofit.create(CharacterService.class);
        }
        return characterService;
    }
    private static void initializeRetrofit(){
        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }
}

File: app/src/main/java/com/ds/rickandmorty/services/CharacterServiceProvider.kt
package com.ds.rickandmorty.services

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object CharacterServiceProvider {

    private const val BASE_URL = "https://rickandmortyapi.com/api/character/"

    private var retrofit: Retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

    private var characterService: CharacterService = retrofit.create(CharacterService::class.java)

    @JvmStatic
    val instanceCharService: CharacterService
        get() = characterService

}